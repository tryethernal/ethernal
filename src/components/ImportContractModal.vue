<template>
<v-dialog v-model="dialog" max-width="600">
    <v-card>
        <v-card-title class="headline">Import a third-party contract</v-card-title>

        <v-card-text>
            <v-alert text v-if="!canImport" type="error">Free plan users are limited to 10 synced contracts. Remove some contracts or <Upgrade-Link @goToBilling="goToBilling" :emit="true">upgrade</Upgrade-Link> to the Premium plan for more.</v-alert>
            <v-alert type="success" v-if="successMessage" v-html="successMessage"></v-alert>
            <v-alert type="error" v-if="errorMessage"> {{ errorMessage }}</v-alert>
            <div v-if="chain.scanner">
                Enter an address of a contract deployed on {{ chain.name }} mainnet.<br>
                If the contract has been verified on {{ chain.scanner }}, its name and ABI will be pulled automatically.<br>
                If not, the contract will be imported but you'll have to manually add the name and ABI.
                To be able to use this, your workspace needs to be connected to a mainnet fork.<br>
                If it is not, the contract will still be imported but calls will fail.<br>
            </div>
            <v-text-field :disabled="!canImport" id="contractAddress" v-model="contractAddress" label="Address*" required></v-text-field>
        </v-card-text>

        <v-card-actions>
            <v-spacer></v-spacer>
            <v-btn color="primary" text @click.stop="close()">Close</v-btn>
            <v-btn id="importContract" color="primary" :loading="loading" :disabled="!contractAddress" text @click.stop="importContract()">Import</v-btn>
        </v-card-actions>
    </v-card>
</v-dialog>
</template>
<script>
import { mapGetters } from 'vuex';
import UpgradeLink from './UpgradeLink';

export default {
    name: 'ImportContractModal',
    components: {
        UpgradeLink
    },
    data: () => ({
        dialog: false,
        resolve: null,
        reject: null,
        contractAddress: null,
        successMessage: null,
        errorMessage: null,
        loading: false,
        options: {}
    }),
    methods: {
        open: function(options) {
            this.dialog = true;
            this.options = options || {};
            return new Promise((resolve, reject) => {
                this.resolve = resolve;
                this.reject = reject;
            });
        },
        close: function() {
            this.resolve(true);
            this.reset();
        },
        goToBilling: function() {
            this.close();
            this.$router.push({ path: '/settings', query: { tab: 'billing' }});
        },
        importContract: function() {
            this.successMessage = null;
            this.errorMessage = null;
            this.loading = true;
            this.server.importContract(this.contractAddress)
                .then(() => {
                        this.successMessage = `Contract has been imported at address <a class="white--text" href="/address/${this.contractAddress}">${this.contractAddress}</a>. If available, metadata will be automatically imported from Etherscan in a few seconds...`
                })
                .catch(error => this.errorMessage = error.message)
                .finally(() => this.loading = false);
        },
        reset: function() {
            this.contractAddress = null;
            this.successMessage = null;
            this.errorMessage = null;
            this.loading = false;
            this.dialog = false;
            this.resolve = null;
            this.reject = null;
        }
    },
    computed: {
        ...mapGetters([
            'currentWorkspace',
            'user',
            'chain'
        ]),
        canImport: function() {
            return this.user.plan == 'premium' || this.options.contractsCount < 10;
        },
    }
}
</script>
