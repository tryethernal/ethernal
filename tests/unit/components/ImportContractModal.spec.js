import MockHelper from '../MockHelper';
  
import ImportContractModal from '@/components/ImportContractModal.vue';

describe('ImportContractModal.vue', () => {
    let helper;

    beforeEach(() => {
        helper = new MockHelper();
    });

    it('Should let the user import a verified mainnet contract', async () => {
        const wrapper = helper.mountFn(ImportContractModal);
        const importContractMock = jest.spyOn(helper.mocks.server, 'importContract');
        await wrapper.setData({ dialog: true, options: { contractsCount: 1 } });

        await wrapper.find('#contractAddress').setValue('0x123');
        await wrapper.find('#importContract').trigger('click');

        await new Promise(process.nextTick);

        expect(importContractMock).toHaveBeenCalled();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should let the user import a non verified mainnet contract', async () => {
        jest.spyOn(helper.mocks.server, 'importContract')
            .mockResolvedValue({ data: { success: true, contractIsVerified: false }});

        const wrapper = helper.mountFn(ImportContractModal);
        const importContractMock = jest.spyOn(helper.mocks.server, 'importContract');
        await wrapper.setData({ dialog: true, options: { contractsCount: 1 } });

        await wrapper.find('#contractAddress').setValue('0x123');
        await wrapper.find('#importContract').trigger('click');

        await new Promise(process.nextTick);

        expect(importContractMock).toHaveBeenCalled();
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should warn the user if he is on a free plan and has already 10 contracts', async () => {
        const wrapper = helper.mountFn(ImportContractModal);
        await wrapper.setData({ dialog: true, options: { contractsCount: 10 } });

        await wrapper.find('#contractAddress').setValue('0x123');
        await wrapper.find('#importContract').trigger('click');

        await new Promise(process.nextTick);
        expect(wrapper.html()).toMatchSnapshot();
    });

    it('Should not warn the user if he is on a premium plan and has already 10 contracts', async () => {
        const wrapper = helper.mountFn(ImportContractModal, {
            getters: {
                user: jest.fn().mockReturnValue({ plan: 'premium' })
            }
        });
        await wrapper.setData({ dialog: true, options: { contractsCount: 10 } });

        await wrapper.find('#contractAddress').setValue('0x123');
        await wrapper.find('#importContract').trigger('click');

        await new Promise(process.nextTick);
        expect(wrapper.html()).toMatchSnapshot();
    });
});
