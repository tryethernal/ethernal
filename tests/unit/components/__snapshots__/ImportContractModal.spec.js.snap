// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ImportContractModal.vue Should let the user import a non verified mainnet contract 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Import a third-party contract</div>
          <div class="v-card__text">
            <!---->
            <div role="alert" class="v-alert v-sheet theme--dark success">Contract has been imported at address <a class="white--text" href="/address/0x123">0x123</a>. If available, metadata will be automatically imported from Etherscan in a few seconds...</div>
            <!---->
            <div>
              Enter an address of a contract deployed on Ethereum mainnet.<br>
              If the contract has been verified on Etherscan, its name and ABI will be pulled automatically.<br>
              If not, the contract will be imported but you'll have to manually add the name and ABI.
              To be able to use this, your workspace needs to be connected to a mainnet fork.<br>
              If it is not, the contract will still be imported but calls will fail.<br></div>
            <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="contractAddress" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Address*</label><input required="required" id="contractAddress" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text" id="importContract"><span class="v-btn__content">Import</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`ImportContractModal.vue Should let the user import a verified mainnet contract 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Import a third-party contract</div>
          <div class="v-card__text">
            <!---->
            <div role="alert" class="v-alert v-sheet theme--dark success">Contract has been imported at address <a class="white--text" href="/address/0x123">0x123</a>. If available, metadata will be automatically imported from Etherscan in a few seconds...</div>
            <!---->
            <div>
              Enter an address of a contract deployed on Ethereum mainnet.<br>
              If the contract has been verified on Etherscan, its name and ABI will be pulled automatically.<br>
              If not, the contract will be imported but you'll have to manually add the name and ABI.
              To be able to use this, your workspace needs to be connected to a mainnet fork.<br>
              If it is not, the contract will still be imported but calls will fail.<br></div>
            <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="contractAddress" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Address*</label><input required="required" id="contractAddress" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text" id="importContract"><span class="v-btn__content">Import</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`ImportContractModal.vue Should not warn the user if he is on a premium plan and has already 10 contracts 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Import a third-party contract</div>
          <div class="v-card__text">
            <!---->
            <div role="alert" class="v-alert v-sheet theme--dark success">Contract has been imported at address <a class="white--text" href="/address/0x123">0x123</a>. If available, metadata will be automatically imported from Etherscan in a few seconds...</div>
            <!---->
            <div>
              Enter an address of a contract deployed on Ethereum mainnet.<br>
              If the contract has been verified on Etherscan, its name and ABI will be pulled automatically.<br>
              If not, the contract will be imported but you'll have to manually add the name and ABI.
              To be able to use this, your workspace needs to be connected to a mainnet fork.<br>
              If it is not, the contract will still be imported but calls will fail.<br></div>
            <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="contractAddress" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Address*</label><input required="required" id="contractAddress" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text" id="importContract"><span class="v-btn__content">Import</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;

exports[`ImportContractModal.vue Should warn the user if he is on a free plan and has already 10 contracts 1`] = `
<div class="v-dialog__container">
  <div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
    <transition-stub name="dialog-transition" appear="true">
      <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; max-width: 600px;">
        <div class="v-card v-sheet theme--light">
          <div class="v-card__title headline">Import a third-party contract</div>
          <div class="v-card__text">
            <div role="alert" class="v-alert v-sheet theme--dark v-alert--text error--text">
              <div class="v-alert__wrapper"><i aria-hidden="true" class="v-icon notranslate v-alert__icon mdi mdi-alert theme--dark error--text"></i>
                <div class="v-alert__content">Free plan users are limited to 10 synced contracts. Remove some contracts or <a>upgrade</a> to the Premium plan for more.</div>
              </div>
            </div>
            <!---->
            <!---->
            <div>
              Enter an address of a contract deployed on Ethereum mainnet.<br>
              If the contract has been verified on Etherscan, its name and ABI will be pulled automatically.<br>
              If not, the contract will be imported but you'll have to manually add the name and ABI.
              To be able to use this, your workspace needs to be connected to a mainnet fork.<br>
              If it is not, the contract will still be imported but calls will fail.<br></div>
            <div class="v-input v-input--is-disabled theme--light v-text-field">
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="contractAddress" class="v-label v-label--is-disabled theme--light" style="left: 0px; position: absolute;">Address*</label><input required="required" disabled="disabled" id="contractAddress" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div class="v-card__actions">
            <div class="spacer"></div> <button type="button" class="v-btn v-btn--text theme--light v-size--default primary--text"><span class="v-btn__content">Close</span></button> <button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--text theme--light v-size--default" id="importContract"><span class="v-btn__content">Import</span></button>
          </div>
        </div>
      </div>
    </transition-stub>
  </div>
</div>
`;
